redfish/data_source_redfish_storage_volume.go

```diff
@@ -0,0 +1,119 @@
+		"redfish_server": {
+			Type:        schema.TypeList,
+			Required:    true,
+			Description: "List of server BMCs and their respective user credentials",
+			Elem: &schema.Resource{
+				Schema: map[string]*schema.Schema{
+					"user": {
+						Type:        schema.TypeString,
+						Optional:    true,
+						Description: "User name for login",
+					},
+					"password": {
+						Type:        schema.TypeString,
+						Optional:    true,
+						Description: "User password for login",
+						Sensitive:   true,
+					},
+					"endpoint": {
+						Type:        schema.TypeString,
+						Required:    true,
+						Description: "Server BMC IP address or hostname",
+					},
+					"ssl_insecure": {
+						Type:        schema.TypeBool,
+						Optional:    true,
+						Description: "This field indicates whether the SSL/TLS certificate must be verified or not",
+					},
+				},
+			},
+		},
+		"storage_volume": {
+			Type:        schema.TypeList,
+			Description: "List of storage volumes available on this instance",
+			Computed:    true,
+			Elem: &schema.Resource{
+				Schema: map[string]*schema.Schema{
+					"storage_controller_id": {
+						Description: "Disks attached to the storage resource",
+						Computed:    true,
+						Type:        schema.TypeString,
+					},
+					"volume_disks": {
+						Type:        schema.TypeList,
+						Description: "Disks attached to the storage resource",
+						Computed:    true,
+						Elem: &schema.Schema{
+							Type: schema.TypeString,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func dataSourceRedfishStorageVolumeRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	service, err := NewConfig(m.(*schema.ResourceData), d)
+	if err != nil {
+		return diag.Errorf(err.Error())
+	}
+	return readRedfishStorageVolumeCollection(service, d)
+}
+
+func readRedfishStorageVolumeCollection(service *gofish.Service, d *schema.ResourceData) diag.Diagnostics {
+	var diags diag.Diagnostics
+	m := make([]map[string]interface{}, 0) //List where all storage controller will be held
+
+	systems, err := service.Systems()
+	if err != nil {
+		return diag.Errorf("Error when retrieving systems: %s", err)
+	}
+	storage, err := systems[0].Storage()
+	if err != nil {
+		return diag.Errorf("Error when retrieving storage: %s", err)
+	}
+
+	var mToAdd map[string]interface{} //Map where each controller and its disks will be written
+	for _, s := range storage {
+		mToAdd = make(map[string]interface{}) //Create new mToAdd instace
+		mToAdd["storage_controller_id"] = s.ID
+		drives, err := s.Drives()
+		if err != nil {
+			return diag.Errorf("Error when retrieving drives: %s", err)
+		}
+
+		driveNames := make([]interface{}, 0)
+		for _, d := range drives {
+			driveNames = append(driveNames, d.Name)
+		}
+		mToAdd["volume_disks"] = driveNames
+		m = append(m, mToAdd) //Insert controller into list
+	}
+
+	d.Set("storage_volume", m)
+	d.SetId(strconv.FormatInt(time.Now().Unix(), 10))
```

Comment:
Yeah, I've used that because in the terraform example datasource uses that for setting the ID.

To keep it consistent makes sense, but from the point of view of usability, it doesn't matter since it is just a datasource.

I'll change it.

Reply:
@anupamaloke 
We cannot use */redfish/v1/Systems/System.Embedded.1/Storage* as ID because the Storage() method from System returns a list of Storage objects and there is no way to get ODataID from the storage collection. 

Either we set that hardcoded, which I would not do, or we keep it like that, with a simple timestamp.

Let me know your thoughts.

Reply:
@mikeletux, sorry for tardiness on my part. Since this data source gets a list of the physical drives attached to a storage controller, one option would be to use the ODataID for that storage controller for e.g. /redfish/v1/Systems/System.Embedded.1/Storage/StorageController/RAID.Slot.1-1 (as I don't have access to an iDRAC right now, this might not be the correct ODataID. However, I just wanted to provide an example). Let me know what you think.

Reply:
Hey Anupam,

This data source does get every storage controller as well as every disk attached to each one. Take a look to the output below:

~~~
  "spd-demo10-idrac" = {
    "id" = "1617091209"
    "redfish_server" = [
      {
        "endpoint" = "XXXXXXXXXXX"
        "password" = "XXXXXXXXXXXX"
        "ssl_insecure" = true
        "user" = "XXXXXXXXXXXXXXXXXX"
      },
    ]
    "storage" = [
      {
        "drives" = [
          "Solid State Disk 0:1:0",
          "Solid State Disk 0:1:1",
        ]
        "storage_controller_id" = "RAID.Integrated.1-1"
      },
      {
        "drives" = []
        "storage_controller_id" = "AHCI.Embedded.3-1"
      },
    ]
  }
~~~

See how we retrieve all disk controllers. In this case we have:
- **RAID.Integrated.1-1** controller with 2 disks attached (*Solid State Disk 0:1:0* and *Solid State Disk 0:1:1*)
- **AHCI.Embedded.3-1** controller with no disks whatsoever

Since we are retrieving all info from storage and not just one, I propose to keep using a timestamp as ID. Also here we are not dealing with a resource, but datasource.

Let me know.

Reply:
@mikeletux, thanks for providing the details. I guess we can go ahead with your current implementation. Let's create a new enhancement issue for the ID computation so that we have an issue available to revisit this and address it holistically later.